src/ir/proto_to_ir.h

```diff
@@ -0,0 +1,122 @@
+//-----------------------------------------------------------------------------
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//----------------------------------------------------------------------------
+#ifndef SRC_IR_PROTO_TO_IR_H_
+#define SRC_IR_PROTO_TO_IR_H_
+
+#include "src/ir/attributes/attribute.h"
+#include "src/ir/block_builder.h"
+#include "src/ir/ir_context.h"
+#include "src/ir/module.h"
+#include "src/ir/proto/raksha_ir.pb.h"
+#include "src/ir/value.h"
+#include "src/ir/ssa_names.h"
+
+namespace raksha::ir {
+
+using ResultType =
+    std::variant<std::unique_ptr<Block>, std::unique_ptr<Operation>, Value,
+                 value::Any, value::BlockArgument, value::OperationResult,
+                 Attribute>;
+
+class ProtoToIR {
+ public:
+  using ID = uint64_t;
+  struct Result {
+    std::unique_ptr<Module> module;
+    std::unique_ptr<SsaNames> ssa_names;
+  };
+
+  ProtoToIR(IRContext& context, SsaNames& ssa_names) : context_{context}, ssa_names_{ssa_names} {}
+
+  // Disable copy (and move) semantics.
+  ProtoToIR(const ProtoToIR&) = delete;
+  ProtoToIR& operator=(const ProtoToIR&) = delete;
+
+  virtual ~ProtoToIR() {}
+
+  static Result Convert(IRContext& context,
+                        const proto::IrTranslationUnit& root) {
+    // Store the operators in the context.
+    for (const auto& op : root.operators()) {
+      auto name = op.name();
+      // Register new operators.
+      if (!context.IsRegisteredOperator(name)) {
+        Operator op = Operator(name);
+        context.RegisterOperator(std::make_unique<Operator>(std::move(op)));
+      }
+    }
+
+    auto ssa_names = std::make_unique<SsaNames>();
+    ProtoToIR state(context, *ssa_names);
+    // Construct objects to point to during construction.
+    auto module = state.PreVisit(root.top_level_module());
+    // Perform construction.
+    state.ToIR(root.top_level_module());
+    // Move objects into their 'owning' objects.
+    state.PostVisit(*module, root.top_level_module());
+
+    // TODO: Check that the construction correctly handled ownership of all
```

Comment:
File an issue?

Reply:
Removed stale comment (following code does the check).